src/Microsoft.AspNet.Mvc.Razor/Precompilation/PrecompilationTagHelperTypeResolver.cs

```diff
@@ -0,0 +1,138 @@
+    public class PrecompilationTagHelperTypeResolver : TagHelperTypeResolver
+    {
+        private readonly object _assemblyLookupLock = new object();
+        private readonly Dictionary<string, IEnumerable<ITypeInfo>> _assemblyLookup
+            = new Dictionary<string, IEnumerable<ITypeInfo>>(StringComparer.OrdinalIgnoreCase);
+        private readonly CodeAnalysis.Compilation _compilation;
+        private INamedTypeSymbol _openGenericDictionaryTypeSymbol;
+
+        public PrecompilationTagHelperTypeResolver([NotNull] CodeAnalysis.Compilation compilation)
+        {
+            _compilation = compilation;
+        }
+
+        /// <inheritdoc />
+        protected override IEnumerable<ITypeInfo> GetExportedTypes([NotNull] AssemblyName assemblyName)
+        {
+            lock (_assemblyLookup)
+            {
+                IEnumerable<ITypeInfo> result;
+                if (!_assemblyLookup.TryGetValue(assemblyName.Name, out result))
+                {
+                    result = GetExportedTypes(assemblyName.Name);
+                    _assemblyLookup[assemblyName.Name] = result;
+                }
+
+                return result;
+            }
+        }
+
+        private INamedTypeSymbol OpenGenericDictionaryType
+        {
+            get
+            {
+                if (_openGenericDictionaryTypeSymbol == null)
+                {
+                    _openGenericDictionaryTypeSymbol = _compilation.GetTypeByMetadataName(
+                        typeof(IDictionary<,>).FullName);
+
+                    Debug.Assert(_openGenericDictionaryTypeSymbol != null);
+                }
+
+                return _openGenericDictionaryTypeSymbol;
+            }
+        }
+
+        // Internal for unit testing
+        internal IEnumerable<ITypeInfo> GetExportedTypes(string assemblyName)
+        {
+            if (string.Equals(_compilation.AssemblyName, assemblyName, StringComparison.Ordinal))
+            {
+                return GetExportedTypes(_compilation.Assembly);
+            }
+            else
+            {
+                foreach (var reference in _compilation.References)
+                {
+                    var compilationReference = reference as CompilationReference;
+                    if (compilationReference != null &&
+                        string.Equals(
+                            compilationReference.Compilation.AssemblyName,
+                            assemblyName,
+                            StringComparison.Ordinal))
+                    {
+                        return GetExportedTypes(compilationReference.Compilation.Assembly);
+                    }
+
+                    var assemblyReference = reference as PortableExecutableReference;
+                    if (assemblyReference != null)
+                    {
+                        var assemblyOrModuleSymbol = _compilation.GetAssemblyOrModuleSymbol(reference);
+                        if (assemblyOrModuleSymbol != null &&
```

Comment:
I haven't seen a module being returned with the set of samples we have. Some of this needs hardening (especially around assemblies\types with errors)